From bc02908bf56a41d9e736370385c50b245c374d82 Mon Sep 17 00:00:00 2001
From: Marek 'marx' Grac <mgrac@redhat.com>
Date: Wed, 20 Mar 2013 10:04:57 +0100
Subject: [PATCH 3/5] Resolves: rhbz#875148 crash with striped segments in
 mirror

Application is fixed so it starts properly. Logical volumes are visible but
volume group information is not shown properly, functionality is not limited.
---
 src/Segment.py  | 10 ++++++++--
 src/renderer.py | 34 ++++++++++++++++++++++------------
 2 files changed, 30 insertions(+), 14 deletions(-)

diff --git a/src/Segment.py b/src/Segment.py
index 0ec56cd..d50c3b8 100644
--- a/src/Segment.py
+++ b/src/Segment.py
@@ -105,8 +105,14 @@ class MirroredSegment(Segment):
     self.image_lvs.append(image_lv)
     # set annotations
     for seg in image_lv.get_segments():
-      extent = seg.get_extent_block()
-      extent.set_annotation(_("Mirror") + str(image_num))
+      extent = None
+      
+      if seg.type != STRIPED_SEGMENT_ID:
+        extent = seg.get_extent_block()
+        extent.set_annotation(_("Mirror") + str(image_num))
+      else:
+        # stripped segments only because they do not map to extents directly
+        pass
   def get_images(self):
     return self.image_lvs
   def clear_images(self):
diff --git a/src/renderer.py b/src/renderer.py
index ed2687e..5974f15 100644
--- a/src/renderer.py
+++ b/src/renderer.py
@@ -354,14 +354,18 @@ class DisplayView:
                 image_lv_cyls = []
                 for image_lv in lv.get_segments()[0].get_images():
                     image_lv_cyl = Subcylinder(self.lv_cyl_gen, 1, 0, False)
-                    label_mirror = "<span size=\"8000\">" + image_lv.get_segments()[0].get_extent_block().get_annotation() + "</span>"
+                    try:
+                        label_mirror = "<span size=\"8000\">" + image_lv.get_segments()[0].get_extent_block().get_annotation() + "</span>"
+                    except:
+                        label_mirror = "<span size=\"8000\"> None </span>"
                     image_lv_cyl.set_label_lower(label_mirror, False, True, False)
                     image_lv_cyls.append(image_lv_cyl)
                     for seg in image_lv.get_segments():
                         # mirror should have linear mapping only
-                        extent = seg.get_extent_block()
-                        subcyl = Subcylinder(self.lv_cyl_gen, 1, 0, False, extent.get_start_size()[1])
-                        image_lv_cyl.children.append(subcyl)
+                        if seg.type == LINEAR_SEGMENT_ID:
+                            extent = seg.get_extent_block()
+                            subcyl = Subcylinder(self.lv_cyl_gen, 1, 0, False, extent.get_start_size()[1])
+                            image_lv_cyl.children.append(subcyl)
                 if len(image_lv_cyls) != 0:
                     cyl.children.append(image_lv_cyls[0])
                 for image_lv_cyl in image_lv_cyls[1:]:
@@ -452,16 +456,20 @@ class DisplayView:
                     image_lv_cyls = []
                     for image_lv in seg.get_images():
                         image_lv_cyl = Subcylinder(self.lv_cyl_gen, 1, 0, False)
-                        label_mirror = "<span size=\"8000\">" + image_lv.get_segments()[0].get_extent_block().get_annotation() + "</span>"
+                        try:
+                            label_mirror = "<span size=\"8000\">" + image_lv.get_segments()[0].get_extent_block().get_annotation() + "</span>"
+                        except:
+                            label_mirror = "<span size=\"8000\">Extent Information Is Not Available</span>"
                         image_lv_cyl.set_label_lower(label_mirror, False, True, False)
                         image_lv_cyls.append(image_lv_cyl)
                         lv_cyls_dir[image_lv] = image_lv_cyl
                         for seg2 in image_lv.get_segments():
                             # mirror should have linear mapping only
-                            extent = seg2.get_extent_block()
-                            subcyl = Subcylinder(self.lv_cyl_gen, 1, 0, False, extent.get_start_size()[1])
-                            lv_cyls_dir[extent] = subcyl
-                            image_lv_cyl.children.append(subcyl)
+                            if seg2.type == LINEAR_SEGMENT_ID:
+                                extent = seg2.get_extent_block()
+                                subcyl = Subcylinder(self.lv_cyl_gen, 1, 0, False, extent.get_start_size()[1])
+                                lv_cyls_dir[extent] = subcyl
+                                image_lv_cyl.children.append(subcyl)
                     if len(image_lv_cyls) != 0:
                         cyl.children.append(image_lv_cyls[0])
                     for image_lv_cyl in image_lv_cyls[1:]:
@@ -522,9 +530,11 @@ class DisplayView:
                 label = label + str(width) + ' extents'
                 label = label + "</span>"
                 ext_cyl_p.set_label_lower(label, False, False, True)
-                ext_cyl_l = lv_cyls_dir[ext]
-                ext_cyl_l.add_highlightable(ext_cyl_p)
-                pv_cyl.children.append(ext_cyl_p)
+                ext_cyl_l = None
+                if lv_cyls_dir.has_key(ext) == True:
+                    ext_cyl_l = lv_cyls_dir[ext]
+                    ext_cyl_l.add_highlightable(ext_cyl_p)
+                    pv_cyl.children.append(ext_cyl_p)
         
         self.display.append_right(True, End(self.lv_cyl_gen))
         for lv_cyl in lv_cyls:
-- 
2.1.4

